tests/src/tests/test_utils.go

```diff
@@ -0,0 +1,394 @@
+	setRandSeed()
+	return rand.Intn(success+failure) > failure-1
+}
+
+// Randomly returns an integer between 1 and the number specified.
+func GetRandomNumber(max int) int {
+	setRandSeed()
+	return rand.Intn(max) + 1
+}
+
+// Return a structure used to represent a specific devfile in the tests
+func GetDevfile(fileName string) TestDevfile {
+	testDevfile := TestDevfile{}
+	testDevfile.SchemaDevFile = schema.Devfile{}
+	testDevfile.FileName = fileName
+	testDevfile.SchemaDevFile.SchemaVersion = "2.0.0"
+	testDevfile.SchemaParsed = false
+	return testDevfile
+}
+
+// Create a devifle on disk for use in the tests.
+// If useParser is true the parser is used to generate the file, otherwise "sigs.k8s.io/yaml" is used.
+// The TestDevfile structure specified contains the name of the devfile and its required content.
+func (devfile *TestDevfile) CreateDevfile(useParser bool) error {
+	var err error
+
+	fileName := devfile.FileName
+	if !strings.HasSuffix(fileName, ".yaml") {
+		fileName += ".yaml"
+	}
+
+	if useParser {
+		LogInfoMessage(fmt.Sprintf("Use Parser to write devfile %s", fileName))
+		newDevfile, err := devfileData.NewDevfileData(devfile.SchemaDevFile.SchemaVersion)
+		if err != nil {
+			LogErrorMessage(fmt.Sprintf("Creating new devfile : %v", err))
+		} else {
+			newDevfile.SetSchemaVersion(devfile.SchemaDevFile.SchemaVersion)
+
+			// add the commands to new devfile
+			for _, command := range devfile.SchemaDevFile.Commands {
+				newDevfile.AddCommands(command)
+			}
+			// add components to the new devfile
+			newDevfile.AddComponents(devfile.SchemaDevFile.Components)
+
+			ctx := devfileCtx.NewDevfileCtx(fileName)
+
+			err = ctx.SetAbsPath()
+			if err != nil {
+				LogErrorMessage(fmt.Sprintf("Setting devfile path : %v", err))
+			} else {
+				devObj := parser.DevfileObj{
+					Ctx:  ctx,
+					Data: newDevfile,
+				}
+				err = devObj.WriteYamlDevfile()
+				if err != nil {
+					LogErrorMessage(fmt.Sprintf("Writing devfile : %v", err))
+				} else {
+					devfile.SchemaParsed = false
+				}
+			}
+
+		}
+	} else {
+		LogInfoMessage(fmt.Sprintf("Marshall and write devfile %s", devfile.FileName))
+		c, err := yaml.Marshal(&(devfile.SchemaDevFile))
+
+		if err != nil {
+			LogErrorMessage(fmt.Sprintf("Marshall devfile %s : %v", devfile.FileName, err))
+		} else {
+			err = ioutil.WriteFile(fileName, c, 0644)
+			if err != nil {
+				LogErrorMessage(fmt.Sprintf("Write devfile %s : %v", devfile.FileName, err))
+			} else {
+				devfile.SchemaParsed = false
+			}
+		}
+	}
+	return err
+}
+
+// Use the parser to parse a devfile on disk
+func (devfile *TestDevfile) ParseSchema() error {
+
+	var err error
+	if !devfile.SchemaParsed {
+		LogInfoMessage(fmt.Sprintf("Parse and Validate %s : ", devfile.FileName))
+		devfile.ParsedSchemaObj, err = devfilepkg.ParseAndValidate(devfile.FileName)
+		if err != nil {
+			LogErrorMessage(fmt.Sprintf("From ParseAndValidate %v : ", err))
+		}
+		devfile.SchemaParsed = true
+	}
```

Comment:
Not sure what the point of parsing is, when you're curating the devfile struct with fake/dummy data in `CreateDevfile`. So the flow basically is populate the devfile struct, write to file and unmarshall it back to structs in a diff variable.. I think this all can be avoided!

Reply:
Is this because you want to verify across two structs in the end?

Reply:
It can look like madness sometimes but when it comes to tests "that is the way".  
The concept is i create a schema structure with attributes which is now my reference.
I write that as a yaml two ways: using the yaml tool and the parser.
I then use the parser to read the the file, and then check the content the parser returns to me matches my reference content.  Under the covers the parse may do "stuff" which sometimes makes this redundant but  "that is the way" and at least the "stuff" is tested. 
Hope you like the Mandalorian humour :-).